### PR TITLE
drivers: spi: Add xmc4xxx driver

### PR DESCRIPTION
**Pull request by talih0**
Adds spi driver for xmc4xxx SoCs.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

### CODE REVIEW

**Comment by talih0** [comment]
image

File: dts/bindings/spi/infineon,xmc4xxx-spi.yaml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+compatible: "infineon,xmc4xxx-spi"
+
+include: [spi-controller.yaml, pinctrl-device.yaml]
+
+properties:
+    reg:
+      required: true
+
+    miso-src:
```
Comment (galak) [inline dts/bindings/spi/infineon,xmc4xxx-spi.yaml:14]: Can this not be handled via pinctrl?

Comment (talih0) [reply dts/bindings/spi/infineon,xmc4xxx-spi.yaml:14]: For xmc4xxx, the input multiplexers are on the peripheral side, and the output muxes are on the GPIO.
In the xmc pinctrl design it is only taking care of the output side. It didn't feel right to have peripheral driver code in pinctrl.

Initially, I thought about adding miso-src (or other input mux info) to the pinctrl dts node, and instead use this information in the peripheral driver, but this was not accepted in the pinctrl review
https://github.com/zephyrproject-rtos/zephyr/pull/47581#discussion_r917667843

**Comment by talih0** [comment]
I rebased to get a fresh CI run. Ping for review pls.

**Comment by talih0** [comment]
ping

**Comment by talih0** [comment]
@tbursztyka, could you have a look at this PR pls?

**Comment by talih0** [comment]
rebased and added note that spi is supported on xmc45_relax_kit board

**Comment by carlescufi** [comment]
@parthitce could you please review?

**Comment by talih0** [comment]
ping

**Comment by talih0** [comment]
ping

**Comment by talih0** [comment]
I rebased to fix merge conflicts

**Comment by talih0** [comment]
ping

**Review by ifyall — DISMISSED** [review]
Lots of good use of XMCLib for the SPI operations. This PR looks good for me.

**Comment by talih0** [comment]
I've rebased to fix merge conflicts.
@ifyall , could you re-ack please? are you no longer in the zephyr-rtos github org?

File: drivers/spi/spi_xmc4xxx.c [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+	XMC_SPI_CH_Transmit(config->spi, tx_data, XMC_SPI_CH_MODE_STANDARD);
+
+	spi_context_update_tx(ctx, 1, 1);
+
+#if defined(CONFIG_SPI_XMC4XXX_INTERRUPT)
+	return;
+#endif
+
+	/* Wait to finish transmitting */
+	while (1) {
```
Comment (ifyall) [inline drivers/spi/spi_xmc4xxx.c:61]: Is there anyway to do this other than an infinite loop? Can we setup an RTOS event or semaphore to let us know when the transmission is complete?

Same comment would apply to the receiving loop on line 69.

Comment (talih0) [reply drivers/spi/spi_xmc4xxx.c:61]: @ifyall, yes, the interrupt driven version returns in line 57 without ever reaching this code. The while (1) {...} loops are only used by the polling version.

Comment (ifyall) [reply drivers/spi/spi_xmc4xxx.c:61]: Thanks. Given that, I have approve.

Comment (talih0) [reply drivers/spi/spi_xmc4xxx.c:61]: > Thanks. Given that, I have approve.

Thanks @ifyall , could you Ack on the PR please

**Comment by talih0** [comment]
I rebased to fix merge conflicts

**Comment by talih0** [comment]
ping

**Review by ifyall — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by teburd — APPROVED** [review]

**Comment by talih0** [comment]
Thank you everyone for reviewing and merging the PR.

**Pull request merged**